### PR TITLE
internal/v4: fix meta endpoints

### PR DESCRIPTION
**Pull request by rogpeppe**
We were printing duplicate endpoints in the /meta list.
Also, when retrieving /meta/any with all meta endpoints included,
it was failing because color was not recognised.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/421/

File: internal/router/router.go [hunk 1]
```diff
@@ -362,7 +362,13 @@ func isNull(val interface{}) bool {
 func (r *Router) metaNames() []string {
 	names := make([]string, 0, len(r.handlers.Meta))
 	for name := range r.handlers.Meta {
-		names = append(names, strings.TrimSuffix(name, "/"))
+		// Ensure that we don't generate duplicate entries
```
Comment (frankban) [inline internal/router/router.go:365]: Heh, this was tricky.

File: internal/v4/api_test.go [hunk 2]
```diff
@@ -280,6 +280,33 @@ func (s *APISuite) TestEndpointGet(c *gc.C) {
...
+	rec := storetesting.DoRequest(c, storetesting.DoRequestParams{
+		Handler: s.srv,
+		URL:     storeURL("precise/wordpress-23/meta"),
+	})
+	var list []string
+	err := json.Unmarshal(rec.Body.Bytes(), &list)
+	c.Assert(err, gc.IsNil)
+
+	listNames := make(map[string]bool)
+	for _, name := range list {
```
Comment (frankban) [inline internal/v4/api_test.go:295]: Is this a check to ensure list is actually a set? Can you add a comment?

**Comment by frankban** [comment]
:+1: with a minor request.
Thank you!

File: internal/v4/api_test.go (hunk 2 above)
Comment (rogpeppe) [inline internal/v4/api_test.go:295]: The assert is for that (sanity check, despite the test in router), but the
loop is to build up the names to check that they're all tested.
Will add a comment.

File: internal/router/router.go [hunk 3]
```diff
@@ -362,7 +362,13 @@ func isNull(val interface{}) bool {
...
+		// when there's an entry for both "x" and "x/".
+		trimmed := strings.TrimSuffix(name, "/")
+		if trimmed != name && r.handlers.Meta[trimmed] != nil {
+			continue
+		}
+		names = append(names, trimmed)
```
Comment (urosj) [inline internal/router/router.go:371]: Good catch.

**Comment by urosj** [comment]
:+1: Thanks.

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**